lib/sax.js

```diff
@@ -581,11 +593,27 @@ function parseDCC(source, start, domBuilder, errorHandler) {
 						sysid = matchs[3][0];
 					}
 				}
-				var lastMatch = matchs[len - 1];
-				domBuilder.startDTD(name, pubid, sysid);
+
+				// Unquote pubid and sysid if necessary
+				if (pubid && /^(['"]).*\1$/.test(pubid)) {
+					pubid = pubid.substring(1, pubid.length - 1);
+				}
+				if (sysid && /^(['"]).*\1$/.test(sysid)) {
+					sysid = sysid.substring(1, sysid.length - 1);
+				}
+
+				// Extract internal subset
+				var doctype = extractDoctype(source, start);
```

Comment:
When calling this code the context is already restricted to  we are inside a doctype, so it can not be a comment or anything else (see condition in line 583)

So to just get the internalSubset I think way less work is required as part of extractDocType. I would have expected it to be between the first `[` after the ids and the last `]`?

THe general approach I prefer is to first get something "minimal" working (including all required tests) and iterate over it in the next steps/PRs.
Especially "optimizations" should not make the code less readable/maintainable.

Reply:
I understand. At the time I wrote that my end goal was replacing the entire part where the doctype is handled. 

I didn't just extract everything from the first `[` to the last `]` because at that point it is not known at which index the doctype declaration ends (or it it?)

I have been experimenting with regex to extract all the contents of DOCTYPE (to replace the regex in `split`) and came up with [this](https://regex101.com/r/fqt0AM/2):

`<!DOCTYPE\s+(\w+)(\s+(SYSTEM)\s+(['"])(\w+)\4|\s+(PUBLIC)\s+(['"])(\w+)\7\s+(['"])(\w+)\9)?\s*(\[(\s*(%\w+;|<([^>]*)>)\s*)*])?\s*>`

then we could get
name: group[1]
publicId: group[8]
systemId: group[10] || group[5]
internalSubset: group[11]

Obviously occurrences of `\w` would have to be replaced with namecharstart+namechar*

Do you think this is worth pursuing further?

Reply:
Yes, I understood that. And if you plan to also implement more things then just preserving the internalSubset, like parsing entities in this PR, using an extra method like this might make sense.
But if this PR just tries to preserve the `internlSubset` (which it already does), we should find the simplest possible implementation and elaborate on it in the next step/PR.

Reply:
> I have been experimenting with regex to extract all the contents of DOCTYPE (to replace the regex in `split`)

I did not yet look at that in every detail yet, but I will, just needs more time.

Reply:
If you decide the regex approach is worth a shot I can dedicate more time to making it spec compliant, then the extractDoctype method will not be needed.
I was planning to use this PR as a stepping stone for a entity parser.

Reply:
But if you change the regex the tests will tell us whether anything breaks, so give it a try, we can always revert!

Reply:
@karfau I have pushed the experimental regex parser